### PR TITLE
progressive restoring of trainer state

### LINKED ISSUE #7535: [bug]Resuming From Checkpoint for FP16 failure (Single GPU)
## Please reproduce using the BoringModel
```python
from typing import Dict, Any
from pytorch_lightning.callbacks import LearningRateMonitor, ModelCheckpoint

from torch.utils.data import DataLoader
import logging
import os
import torch
import torch.nn as nn
import torch.optim as optim

import pytorch_lightning as pl

from torch.optim import AdamW

class ToyModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.net1 = nn.Linear(10, 10)
        self.relu = nn.ReLU()
        self.net2 = nn.Linear(10, 5)

    def forward(self, x):
        return self.net2(self.relu(self.net1(x)))

class ToyTask(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.loss_fn = nn.MSELoss()
    
    def setup(self, stage: str):
        if stage == "test":
            return
    
        self.model = ToyModel()
        self.optimizer = AdamW(self.model.parameters(), lr=0.001, betas=[0.9, 0.999], eps=1.0e-08, weight_decay=0, amsgrad=False)

    def forward(self, x):
        return self.model(x)
        
    def training_step(self, batch, batch_idx):
        targets = self.forward(batch["model_input"])
        loss = self.loss_fn(targets, batch["label"]) 

        # Log loss results per train step and per epoch
        self.log("loss", loss)

        # Tell Lightning to minimize loss
        return loss
            
    def configure_optimizers(self):
        return self.optimizer
    
    def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
        self.setup("fit")
```

setup training
```python
task = ToyTask()

dataset = [
    {"model_input": torch.randn(20, 10), "label": torch.randn(20, 5)} for _ in range(10)
]

train_dataloader = DataLoader(dataset, batch_size=None)
val_dataloader = DataLoader(dataset, batch_size=None)

    
model_checkpoint = ModelCheckpoint(
   save_last=True,
   every_n_val_epochs=1,
)

trainer = pl.Trainer(
    gpus=1,
    precision=16,
    max_epochs=3,
    progress_bar_refresh_rate=100,
    log_gpu_memory=None,
    reload_dataloaders_every_epoch=True,
    limit_train_batches=10,
    limit_val_batches=10,
    limit_test_batches=10,
    callbacks=[model_checkpoint],
)

results = trainer.fit(task, train_dataloader)


```

resume from checkpoint

```python
trainer = pl.Trainer(
    gpus=1,
    precision=16,
    max_epochs=4,
    reload_dataloaders_every_epoch=True,
    limit_train_batches=10,
    limit_val_batches=10,
    limit_test_batches=10,
    callbacks=[model_checkpoint],
    resume_from_checkpoint=model_checkpoint.last_model_path,
)
trainer.fit(task, train_dataloader) <--- this is where will fail
```

breaks at the first training step: 
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/precision/native_amp.py#L96

complains about
```
/mnt/xarfuse/uid-26337/a9b1f2c7-seed-nspid4026533638_cgpid18793775-ns-4026533627/torch/cuda/amp/grad_scaler.py in step(self, optimizer, *args, **kwargs)
    334             self.unscale_(optimizer)
    335 
--> 336         assert len(optimizer_state["found_inf_per_device"]) > 0, "No inf checks were recorded for this optimizer."
    337 
    338         retval = self._maybe_opt_step(optimizer, optimizer_state, *args, **kwargs)

AssertionError: No inf checks were recorded for this optimizer.
```

### Expected behavior

Expected to resume training.

### Environment

**Note**: `Bugs with code` are solved faster ! `Colab Notebook` should be made `public` !

* `IDE`: Please, use our python [bug_report_model.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py
) template.

* `Colab Notebook`: Please copy and paste the output from our [environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py) (or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #7535

Includes the changes from #7955, target branch will be changed to master.

The restore flow before:

```python
call to trainer.fit()

model.setup("fit")  # trainer calls setup
call_configure_sharded_model(model)
accelerator.setup(model) 
...

dispatch() #  _run_train, etc.
restore() # restore everything at once (model, callbacks, optimizers, loop state etc.)
```

The restore flow now:

```python
call to Trainer.fit()

# check if/what we needs to resume, load the checkpoint file
resume_start() 

model.setup("fit")  # trainer calls setup

# model weights get restored as soon as model is setup
restore_datamodule() # calls datamodule.on_load_checkpoint()
restore_model() # calls model.on_load_checkpoint()
restore_callbacks() # calls callback.on_load_checkpoint()

call_configure_sharded_model(model)
accelerator.setup(model) 

pre_dispatch()

# restore training state as soon as everything
restore_training_state()  # restore optimizer, precision, loop progress

dispatch() #  _run_train, etc.

resume_end() # delete checkpoint file in memory
```

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

## Follow up to this PR

Make plugin decide if we want to restore model weights before or after call to `configure_sharded_model` hook as per https://github.com/PyTorchLightning/pytorch-lightning/issues/7535#issuecomment-854917054

```python
if TrainingTypePlugin.restore_before_configure_sharded_model:
    restore_model() # also calls model.on_load_checkpoint()

call_configure_sharded_model(model)

if not TrainingTypePlugin.restore_before_configure_sharded_model:
    restore_model() # also calls model.on_load_checkpoint()
```

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 1]
```diff
@@ -144,6 +143,14 @@ def results(self) -> Optional[Union[_EVALUATE_OUTPUT, _PREDICT_OUTPUT]]:
...
         return self._results
 
+    @property
+    def plugin_restores_model(self) -> bool:
+        return False
+
+    @property
+    def plugin_restores_optimizers(self) -> bool:
+        return False
+
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:153]: instead of having split responsibilities between the checkpoint connector and plugin, I'd prefer the accelerator/plugin always owning the responsibilities for saving/loading both model and optimizer states.

- for oncoming items like https://github.com/PyTorchLightning/pytorch-lightning/issues/6691
- for easier debugging/codebase comprehensibility (rather than hopping through plugin settings)

wdyt?

Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:155]: yes, I was experimenting and it doesn't work well with these properties. Looking for a different solution now. 
But I will generally try to push back on bloating the training type plugin with such responsibilities. Currently, there is the deepspeed plugin that is calling the hooks on the LM. If we aim towards a consistent hook call order, we can't have plugins override it and implement arbitrary logic. Also it leads to duplication and I am worried that people will see this pattern and repeat it. 

I'm looking at two possible now hooks in the plugin now:

```python
    def load_model_state_dict(self, checkpoint: Mapping[str, Any]) -> None:
        self.lightning_module.load_state_dict(checkpoint["state_dict"])

    def load_checkpoint_file(self, checkpoint_path: Union[str, Path]) -> Dict[str, Any]:
        return pl_load(checkpoint_path, map_location=(lambda storage, loc: storage))
```
the checkpoint connector would still lhave full control in which order these are called.

Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py]: @ananthsub does this look better to you: https://github.com/PyTorchLightning/pytorch-lightning/pull/7928
(split out changes from this PR)

**Comment by kaushikb11** [comment]
@awaelchli Is the target branch correct?

**Comment by awaelchli** [comment]
will be changed to master because it includes #7955

File: tests/callbacks/test_finetuning_callback.py [hunk 2]
```diff
@@ -275,7 +275,7 @@ def configure_optimizers(self):
     model = FreezeModel()
     cb = OnEpochLayerFinetuning()
     trainer = Trainer(max_epochs=10, resume_from_checkpoint=chk.last_model_path, callbacks=[cb])
-    with pytest.raises(IndexError, match="index 6 is out of range"):
+    with pytest.raises(ValueError, match="loaded state dict has a different number of parameter groups"):
```
Comment (carmocca) [inline tests/callbacks/test_finetuning_callback.py:278]: why has this changed? How is it related to the changes in the PR?

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -849,8 +861,7 @@ def _pre_training_routine(self):
         if self.is_global_zero and self.weights_summary is not None and not self.testing:
             ref_model.summarize(mode=self.weights_summary)
 
-        # restore training and model before hpc is called
-        self.checkpoint_connector.restore()
+        self.checkpoint_connector.resume_end()
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:864]: Why is `resume_end` here, where it runs only if training but `resume_start` in `_run` where it runs for all stages?

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -799,6 +808,9 @@ def _run(self, model: LightningModule) -> Optional[Union[_EVALUATE_OUTPUT, _PRED
     def _pre_dispatch(self):
         self.accelerator.pre_dispatch(self)
 
+        # restore optimizers, etc.
+        self.checkpoint_connector.restore_training_state()
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:812]: Why is this here and not in `_run` as the rest of the `checkpoint_connector` calls?

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 5]
```diff
@@ -357,6 +363,7 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
 
             self.trainer.precision_plugin.on_save_checkpoint(checkpoint)
 
+        # dump hyper-parameters
         # dump hyper-parameters
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:367]:
```suggestion
        # dump hyper-parameters
```

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py]: You are right! `resume_start` should be in `Trainer.fit()`. This was also the reason one of the early stopping tests failed. 
To match the behavior on master, `Trainer(resume_from_checkpoint)` applies only to fit currently.

File: pytorch_lightning/trainer/trainer.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py]: moved to _run() !

File: tests/callbacks/test_finetuning_callback.py (hunk 2 above)
Comment (awaelchli) [reply tests/callbacks/test_finetuning_callback.py:278]: This is because we are now loading the training state, including the optimizer state, earlier in the run phase. 
Previously, this would fail in a IndexError in the on_train_epoch_start of the Finetune callback. Now, it is failing earlier in the run phase when loading the optimizer states. The reason is the same, the layers have changed so the optimizer states can't be loaded 1:1.

File: tests/models/test_hooks.py [hunk 6]
```diff
@@ -500,6 +500,54 @@ def test_trainer_model_hook_system_test(tmpdir):
...
+    )
+    assert called == []
+    trainer.fit(model)
+    expected = [
+        'prepare_data',
+        'configure_callbacks',
+        'setup',
+        'on_load_checkpoint',
+        'configure_sharded_model',
+        'configure_optimizers',
```
Comment (awaelchli) [inline tests/models/test_hooks.py:535]: Hey @carmocca I added a test here using your hook system for the setting 
Trainer(resume_from_checkpoint=..) and this will call on_load_checkpoint() on the model, which I have asserted here to run after setup and before configure_sharded model, which is what we need for the linked issue we try to resolve. 

overall, does this test look ok to you?

Comment (carmocca) [reply tests/models/test_hooks.py:535]: Yes.

IMO you could merge it with `test_trainer_model_hook_system_fit_no_val`.
Do you see a benefit in having them separate?

Comment (awaelchli) [reply tests/models/test_hooks.py:535]: Okay, I just need to save a checkpoint beforehand, in that test so we have something to load.

Comment (awaelchli) [reply tests/models/test_hooks.py]: made the merge, let me know if something is unclear. added comments

**Review by carmocca — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

File: tests/models/test_hooks.py [hunk 7]
```diff
@@ -418,24 +418,39 @@ def test_trainer_model_hook_system_fit(tmpdir):
     assert called == expected
 
 
-def test_trainer_model_hook_system_fit_no_val(tmpdir):
+def test_trainer_model_hook_system_fit_with_resume(tmpdir):
```
Comment (carmocca) [inline tests/models/test_hooks.py:421]:
```suggestion
def test_trainer_model_hook_system_fit_no_val_and_resume(tmpdir):
```

**Review by justusschock — APPROVED** [review]
